Mathlib/Topology/Category/Stonean/Limits.lean

```diff
@@ -129,3 +128,164 @@ Limits.IsColimit.coconePointUniqueUpToIso
+  continuous_toFun := by
+    apply Continuous.subtype_mk
+    exact a.continuous
+
+lemma pullback.condition {X Y Z : Stonean.{u}} (f : X ⟶ Z) {i : Y ⟶ Z}
+    (hi : OpenEmbedding i) : pullback.fst f hi ≫ f = pullback.snd f hi ≫ i := by
+  ext ⟨x, h⟩
+  simp only [Set.mem_preimage] at h
+  obtain ⟨y, hy⟩ := h
+  simp only [fst, snd, comp_apply]
+  change f x = _
+  rw [← hy, @ContinuousMap.coe_mk _ _ (Stonean.instTopologicalSpace (pullback f hi)) _ _ _]
+  congr
+  apply_fun (Homeomorph.ofEmbedding i hi.toEmbedding)
+  simpa only [Homeomorph.ofEmbedding, Homeomorph.homeomorph_mk_coe, Equiv.ofInjective_apply,
+    Homeomorph.homeomorph_mk_coe_symm, Set.MapsTo.restrict, Subtype.map, Function.comp_apply,
+    Equiv.apply_symm_apply, Subtype.mk.injEq]
+
+@[reassoc (attr := simp)]
+lemma pullback.lift_fst {W : Stonean} (a : W ⟶ X) (b : W ⟶ Y) (w : a ≫ f = b ≫ i) :
+  pullback.lift f hi a b w ≫ pullback.fst f hi = a := rfl
+
+@[reassoc (attr := simp)]
+lemma pullback.lift_snd {X Y Z W : Stonean} (f : X ⟶ Z) {i : Y ⟶ Z} (hi : OpenEmbedding i)
+    (a : W ⟶ X) (b : W ⟶ Y) (w : a ≫ f = b ≫ i) :
+    pullback.lift f hi a b w ≫ Stonean.pullback.snd f hi = b := by
+  congr
+  ext z
+  have := congr_fun (FunLike.ext'_iff.mp w.symm) z
+  have h : i (b z) = f (a z) := this
+  suffices : b z = (Homeomorph.ofEmbedding i hi.toEmbedding).symm (⟨f (a z), by rw [← h]; simp⟩)
+  · exact this.symm
+  apply_fun (Homeomorph.ofEmbedding i hi.toEmbedding)
+  simpa only [Homeomorph.ofEmbedding, Homeomorph.homeomorph_mk_coe, Equiv.ofInjective_apply,
+    Homeomorph.homeomorph_mk_coe_symm, Equiv.apply_symm_apply, Subtype.mk.injEq]
+
+/-- The pullback cone whose cone point is the explicit pullback. -/
+@[simps! pt π]
+noncomputable
+def pullback.cone : Limits.PullbackCone f i :=
+  Limits.PullbackCone.mk (pullback.fst f hi) (pullback.snd f hi) (pullback.condition f hi)
+
+lemma pullback.hom_ext {X Y Z W : Stonean} (f : X ⟶ Z) {i : Y ⟶ Z} (hi : OpenEmbedding i)
+    (a : W ⟶ (pullback.cone f hi).pt) (b : W ⟶ (pullback.cone f hi).pt)
+    (hfst : a ≫ pullback.fst f hi = b ≫ pullback.fst f hi) : a = b := by
+  ext z
+  apply_fun (fun q => q z) at hfst
+  apply Subtype.ext
+  exact hfst
+
+/-- The explicit pullback cone is a limit cone. -/
+def pullback.isLimit  : IsLimit (pullback.cone f hi) :=
+  Limits.PullbackCone.isLimitAux _
+    (fun s => pullback.lift f hi s.fst s.snd s.condition)
+    (fun _ => pullback.lift_fst _ _ _ _ _)
+    (fun _ => pullback.lift_snd _ _ _ _ _)
+    (fun _ _ hm => pullback.hom_ext _ _ _ _ (hm .left))
+
+lemma HasPullbackOpenEmbedding : HasPullback f i :=
+  ⟨pullback.cone f hi, pullback.isLimit f hi⟩
+
+section Isos
+
+/-- The isomorphism from the explicit pullback to the abstract pullback. -/
+@[simps]
+noncomputable
+def pullbackIsoPullback : Stonean.pullback f hi ≅
+    @Limits.pullback _ _ _ _ _ f i (HasPullbackOpenEmbedding f hi) where
+  hom :=
+    haveI : HasPullback f i := HasPullbackOpenEmbedding f hi
```

Comment:
It would be nice if we could avoid the repetition of this `haveI`.

Reply:
I changed it to `:=` instead of `where` and then it's enough to state it once